From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Mon, 11 Aug 2025 09:25:39 +0000
Subject: [PATCH 112/245] NVIDIA: SAUCE: tegra_hv: stop ivc ops for unreserved
 IVC queue

till now driver was allowing the IVC ops for before
reserving and after unreserveing the ivc queue.
add checks in ivc APIs to perform ivc ops only on
reserved ivc queues.

Bug 5433066
Bug 5579376

Change-Id: I37553295b1f89dbbdc41b3d23d8fdc53b4f6580b
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3469942
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
(cherry picked from commit 23803b31af948055dc98a6a87d89abe725350a85)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index c2ce088cbe964..9f38c60b1068f 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -926,6 +926,11 @@ void tegra_hv_ivc_notify(struct tegra_hv_ivc_cookie *ivck)
 		return;
 
 	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return;
+	}
+
 	if (WARN_ON(!ivc->cookie.notify_va))
 		return;
 	*ivc->cookie.notify_va = ivc->qd->raise_irq;
@@ -987,7 +992,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_unreserve);
 int tegra_hv_ivc_write(struct tegra_hv_ivc_cookie *ivck, const void *buf,
 		int size)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_write(&ivc->ivc, NULL, buf, size);
 }
@@ -996,7 +1010,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_write);
 int tegra_hv_ivc_write_user(struct tegra_hv_ivc_cookie *ivck, const void __user *buf,
 		int size)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_write(&ivc->ivc, buf, NULL, size);
 }
@@ -1004,7 +1027,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_write_user);
 
 int tegra_hv_ivc_read(struct tegra_hv_ivc_cookie *ivck, void *buf, int size)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_read(&ivc->ivc, NULL, buf, size);
 }
@@ -1012,7 +1044,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_read);
 
 int tegra_hv_ivc_read_user(struct tegra_hv_ivc_cookie *ivck, void __user *buf, int size)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_read(&ivc->ivc, buf, NULL, size);
 }
@@ -1020,7 +1061,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_read_user);
 
 int tegra_hv_ivc_can_read(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_can_read(&ivc->ivc);
 }
@@ -1028,7 +1078,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_can_read);
 
 int tegra_hv_ivc_can_write(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_can_write(&ivc->ivc);
 }
@@ -1036,7 +1095,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_can_write);
 
 int tegra_hv_ivc_tx_empty(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_empty(&ivc->ivc, &ivc->ivc.tx.map);
 }
@@ -1045,7 +1113,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_tx_empty);
 #ifndef CONFIG_KERNEL_BUILD_WITH_PROD_DEFCONFIG
 uint32_t tegra_hv_ivc_tx_frames_available(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return 0;
+	}
 
 	return tegra_ivc_frames_available(&ivc->ivc, &ivc->ivc.tx.map);
 }
@@ -1053,7 +1130,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_tx_frames_available);
 
 int tegra_hv_ivc_dump(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return ivc_dump(ivc);
 }
@@ -1062,10 +1148,19 @@ EXPORT_SYMBOL(tegra_hv_ivc_dump);
 
 void *tegra_hv_ivc_read_get_next_frame(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
 	struct iosys_map map;
 	int err;
 
+	if (ivck == NULL)
+		return ERR_PTR(-EINVAL);
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return ERR_PTR(-EPERM);
+	}
+
 	err = tegra_ivc_read_get_next_frame(&ivc->ivc, &map);
 	if (err)
 		return ERR_PTR(err);
@@ -1075,10 +1170,19 @@ EXPORT_SYMBOL(tegra_hv_ivc_read_get_next_frame);
 
 void *tegra_hv_ivc_write_get_next_frame(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
 	struct iosys_map map;
 	int err;
 
+	if (ivck == NULL)
+		return ERR_PTR(-EINVAL);
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return ERR_PTR(-EPERM);
+	}
+
 	err = tegra_ivc_write_get_next_frame(&ivc->ivc, &map);
 	if (err)
 		return ERR_PTR(err);
@@ -1089,7 +1193,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_write_get_next_frame);
 
 int tegra_hv_ivc_write_advance(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_write_advance(&ivc->ivc);
 }
@@ -1097,7 +1210,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_write_advance);
 
 int tegra_hv_ivc_read_advance(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_read_advance(&ivc->ivc);
 }
@@ -1178,7 +1300,16 @@ EXPORT_SYMBOL(tegra_hv_mempool_unreserve);
 
 int tegra_hv_ivc_channel_notified(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return -EINVAL;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return -EPERM;
+	}
 
 	return tegra_ivc_notified(&ivc->ivc);
 }
@@ -1186,7 +1317,16 @@ EXPORT_SYMBOL(tegra_hv_ivc_channel_notified);
 
 void tegra_hv_ivc_channel_reset(struct tegra_hv_ivc_cookie *ivck)
 {
-	struct hv_ivc *ivc = cookie_to_ivc_dev(ivck);
+	struct hv_ivc *ivc;
+
+	if (ivck == NULL)
+		return;
+
+	ivc = cookie_to_ivc_dev(ivck);
+	if (!ivc->reserved) {
+		pr_err("%s: Error ivc queue is not reserved\n", __func__);
+		return;
+	}
 
 	if (ivc->cookie_ops) {
 		ERR("reset unsupported with callbacks");
-- 
2.43.0

